doc/literatex.1
.\" Automatically generated by Pandoc 3.0.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "LITERATEX" "1" "" "literatex-haskell 0.4.0.0 (2025-01-03)" "literatex Manual"
.nh
.SH NAME
.PP
\f[V]literatex\f[R] - transform literate source code to Markdown
.SH SYNOPSIS
.PP
\f[V]literatex\f[R] [\f[I]OPTIONS\f[R]]
.SH DESCRIPTION
.PP
LiterateX transforms literate source code to Markdown.
Write documentation in Markdown format in the comments of source code,
and LiterateX can transform the file to Markdown, optionally including
the source code with syntax highlighting and line numbers.
.SH OPTIONS
.TP
-h, --help
show help and exit
.TP
--version
show version and exit
.TP
-s, --source \f[I]SOURCE\f[R]
source format
.RS
.PP
The following source formats are supported:
.IP \[bu] 2
\f[V]ddash\f[R] - \f[V]--\f[R] comments
.IP \[bu] 2
\f[V]dslash\f[R] - \f[V]//\f[R] comments
.IP \[bu] 2
\f[V]hash\f[R] - \f[V]#\f[R] comments
.IP \[bu] 2
\f[V]lisp\f[R] - Lisp semicolon comments
.IP \[bu] 2
\f[V]lhs\f[R] - literate Haskell
.IP \[bu] 2
\f[V]percent\f[R] - \f[V]%\f[R] comments
.RE
.TP
-t, --target \f[I]TARGET\f[R]
target format (default: \f[V]pandoc\f[R])
.RS
.PP
The following target formats are supported:
.IP \[bu] 2
\f[V]pandoc\f[R] - Pandoc Markdown
.IP \[bu] 2
\f[V]github\f[R] - GitHub Flavored Markdown
.IP \[bu] 2
\f[V]mdbook\f[R] - mdBook Markdown
.RE
.TP
-l, --language \f[I]LANGUAGE\f[R]
code language
.TP
--ignore-shebang
ignore shebangs
.TP
--no-code
do not display code
.TP
--no-numbers
do not number code lines
.TP
--no-highlight
do not highlight code
.TP
-i, --input \f[I]FILE\f[R]
input file (default: \f[V]STDIN\f[R])
.TP
-o, --output \f[I]FILE\f[R]
output file (default: \f[V]STDOUT\f[R])
.SH Default Options
.TP
\f[V].c\f[R]
c (dslash)
.TP
\f[V].clj\f[R]
clojure (lisp)
.TP
\f[V].css\f[R]
css (dslash)
.TP
\f[V].elm\f[R]
elm (ddash)
.TP
\f[V].erl\f[R]
erlang (percent)
.TP
\f[V].ex\f[R]
elixir (hash)
.TP
\f[V].exs\f[R]
elixir (hash)
.TP
\f[V].go\f[R]
go (dslash)
.TP
\f[V].hs\f[R]
haskell (ddash)
.TP
\f[V].idr\f[R]
idris (ddash)
.TP
\f[V].java\f[R]
java (dslash)
.TP
\f[V].js\f[R]
javascript (dslash)
.TP
\f[V].kt\f[R]
kotlin (dslash)
.TP
\f[V].lhs\f[R]
haskell (lhs)
.TP
\f[V].lisp\f[R]
commonlisp (lisp)
.TP
\f[V].lua\f[R]
lua (ddash)
.TP
\f[V].php\f[R]
php (dslash)
.TP
\f[V].pl\f[R]
perl (hash)
.TP
\f[V].py\f[R]
python (hash)
.TP
\f[V].r\f[R]
r (hash)
.TP
\f[V].rb\f[R]
ruby (hash)
.TP
\f[V].rkt\f[R]
scheme (lisp)
.TP
\f[V].rs\f[R]
rust (dslash)
.TP
\f[V].sc\f[R]
scala (dslash)
.TP
\f[V].scm\f[R]
scheme (lisp)
.TP
\f[V].sh\f[R]
bash (hash)
.TP
\f[V].sql\f[R]
sql (ddash)
.TP
\f[V].tex\f[R]
latex (percent)
.TP
\f[V].ts\f[R]
typescript (dslash)
.SH PROJECT
.TP
GitHub:
<https://github.com/ExtremaIS/literatex-haskell>
.TP
Reporting issues:
GitHub: <https://github.com/ExtremaIS/literatex-haskell/issues>
.RS
.PP
Email: <bugs@extrema.is>
.RE
.TP
Copyright
Copyright (c) 2021-2025 Travis Cardwell
.TP
License
The MIT License <https://opensource.org/licenses/MIT>


CHANGELOG.md
# `literatex-haskell` Changelog

This project follows the [Haskell package versioning policy][PVP], with
versions in `A.B.C.D` format.  `A` may be incremented arbitrarily for
non-technical reasons, but [semantic versioning][SemVer] is otherwise
followed, where `A.B` is the major version, `C` is the minor version, and `D`
is the patch version.  Initial development uses versions `0.0.C.D`, for which
every version is considered breaking.

[PVP]: <https://pvp.haskell.org/>
[SemVer]: <https://semver.org/>

The format of this changelog is based on [Keep a Changelog][KaC], with the
following conventions:

* Level-two heading `Unreleased` is used to track changes that have not been
  released.
* Other level-two headings specify the release in `A.B.C.D (YYYY-MM-DD)`
  format, with newer versions above older versions.
* Level-three headings are used to categorize changes as follows:
    1. Breaking
    2. Non-Breaking
* Changes are listed in arbitrary order and present tense.

[KaC]: <https://keepachangelog.com/en/1.0.0/>

## Unreleased

## 0.4.0.0 (2025-01-03)

### Breaking

* Remove support for GHC 8.6, constraining lower bounds
* Remove support for GHC 8.4, constraining lower bounds
* Remove support for GHC 8.2, constraining lower bounds
* Change minimal Cabal from 1.24 to 3.0

### Non-Breaking

* Bump `base` dependency version upper bound
* Bump `bytestring` dependency version upper bound
* Bump `filepath` dependency version upper bound
* Bump `tasty` dependency version upper bound
* Bump `text` dependency version upper bound
* Bump `ttc` dependency version upper bound

## 0.3.0.0 (2023-05-28)

## Breaking

* Add support for `optparse-applicative` `0.18`

### Non-Breaking

* Bump `ansi-wl-pprint` dependency version upper bound

## 0.2.1.0 (2023-03-21)

### Breaking

* Add `MdBook` target format

### Non-Breaking

* Bump TTC dependency version upper bound
* Adjust dependency constraints to match tested versions

## 0.2.0.2 (2022-02-05)

### Non-Breaking

* Bump `optparse-applicative` dependency version upper bound

## 0.2.0.1 (2021-12-25)

### Non-Breaking

* Bump `text` dependency version upper bound

## 0.2.0.0 (2021-06-25)

### Breaking

* Fix `--help` when using `optparse-applicative` `0.16`

### Non-Breaking

* Refactor Nix configuration
* Use TTC 1.1.0.1

## 0.1.0.2 (2021-06-10)

### Non-Breaking

* Bump TTC dependency version upper bound

## 0.1.0.1 (2021-06-03)

### Non-Breaking

* Use `docker-pkg` scripts to build packages
* Bump TTC dependency version upper bound

## 0.1.0.0 (2021-05-26)

### Breaking

* Initial public release


src/LiterateX.hs
------------------------------------------------------------------------------
-- |
-- Module      : LiterateX
-- Description : API
-- Copyright   : Copyright (c) 2021-2025 Travis Cardwell
-- License     : MIT
--
-- This module provides high-level as well as low-level API functions for
-- transforming literate source code.
------------------------------------------------------------------------------

{-# LANGUAGE FlexibleInstances #-}

module LiterateX
  ( -- * Constants
    version
    -- * API
    -- ** High-Level
    -- $HighLevelAPI
  , transformTextToText
  , transformTextToHandle
  , transformTextToFile
  , transformHandleToText
  , transformHandleToHandle
  , transformHandleToFile
  , transformFileToText
  , transformFileToHandle
  , transformFileToFile
    -- ** Low-Level
    -- $LowLevelAPI
  , runPure
  , runIO
  , runResource
    -- *** Producers
  , sourceString
  , sourceText
  , sourceLazyText
  , sourceByteString
  , sourceLazyByteString
  , sourceHandle
  , sourceFile
    -- *** Consumers
  , sinkString
  , sinkText
  , sinkLazyText
  , sinkByteString
  , sinkLazyByteString
  , sinkHandle
  , sinkFile
    -- *** Transformers
  , transform
  ) where

-- https://hackage.haskell.org/package/base
import Control.Monad.IO.Class (MonadIO)
import Data.Version (showVersion)
import System.IO (Handle)

-- https://hackage.haskell.org/package/bytestring
import qualified Data.ByteString as BS
import qualified Data.ByteString.Lazy as BSL

-- https://hackage.haskell.org/package/conduit
import qualified Conduit as C
import Data.Conduit ((.|))
import qualified Data.Conduit.Combinators as CC
import qualified Data.Conduit.List as CL

-- https://hackage.haskell.org/package/text
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE
import qualified Data.Text.Encoding.Error as TEE
import qualified Data.Text.Lazy as TL

-- https://hackage.haskell.org/package/unliftio
import UnliftIO (MonadUnliftIO)

-- (literatex)
import qualified LiterateX.Parser as Parser
import qualified LiterateX.Renderer as Renderer
import LiterateX.Types (SourceFormat)

-- (literatex:cabal)
import qualified Paths_literatex as Project

------------------------------------------------------------------------------
-- $Constants

-- | LiterateX version string (\"@literatex-haskell X.X.X@\")
--
-- @since 0.0.1.0
version :: String
version = "literatex-haskell " ++ showVersion Project.version

------------------------------------------------------------------------------
-- $HighLevelAPI
--
-- This high-level API provides functions for transforming literate source
-- code.  These functions provide support for transforming from/to lazy
-- 'TL.Text', 'Handle's, and files.
--
-- Example usage:
--
-- @
-- {-\# LANGUAGE OverloadedStrings \#-}
--
-- module Main (main) where
--
-- -- https://hackage.haskell.org/package/literatex
-- import qualified LiterateX
-- import qualified LiterateX.Renderer as Renderer
-- import qualified LiterateX.Types.SourceFormat as SourceFormat
--
-- main :: IO ()
-- main = LiterateX.transformFileToFile
--     SourceFormat.LiterateHaskell
--     (Renderer.defaultOptionsFor "haskell")
--     "demo.lhs"
--     "demo.md"
-- @

-- | Transform from lazy 'TL.Text' to lazy 'TL.Text'
--
-- @since 0.0.1.0
transformTextToText
  :: SourceFormat
  -> Renderer.Options
  -> TL.Text
  -> TL.Text
transformTextToText sourceFormat rendererOpts source =
    runPure sourceFormat rendererOpts (sourceLazyText source) sinkLazyText

-- | Transform from lazy 'TL.Text' to a 'Handle'
--
-- @since 0.0.1.0
transformTextToHandle
  :: MonadIO m
  => SourceFormat
  -> Renderer.Options
  -> TL.Text
  -> Handle
  -> m ()
transformTextToHandle sourceFormat rendererOpts source target =
    runIO sourceFormat rendererOpts
      (sourceLazyText source)
      (sinkHandle target)

-- | Transform from lazy 'TL.Text' to a file
--
-- @since 0.0.1.0
transformTextToFile
  :: MonadUnliftIO m
  => SourceFormat
  -> Renderer.Options
  -> TL.Text
  -> FilePath
  -> m ()
transformTextToFile sourceFormat rendererOpts source target =
    runResource sourceFormat rendererOpts
      (sourceLazyText source)
      (sinkFile target)

-- | Transform from a 'Handle' to lazy 'TL.Text'
--
-- @since 0.0.1.0
transformHandleToText
  :: MonadIO m
  => SourceFormat
  -> Renderer.Options
  -> Handle
  -> m TL.Text
transformHandleToText sourceFormat rendererOpts source =
    runIO sourceFormat rendererOpts (sourceHandle source) sinkLazyText

-- | Transform from a 'Handle' to a 'Handle'
--
-- @since 0.0.1.0
transformHandleToHandle
  :: MonadIO m
  => SourceFormat
  -> Renderer.Options
  -> Handle
  -> Handle
  -> m ()
transformHandleToHandle sourceFormat rendererOpts source target =
    runIO sourceFormat rendererOpts (sourceHandle source) (sinkHandle target)

-- | Transform from a 'Handle' to a file
--
-- @since 0.0.1.0
transformHandleToFile
  :: MonadUnliftIO m
  => SourceFormat
  -> Renderer.Options
  -> Handle
  -> FilePath
  -> m ()
transformHandleToFile sourceFormat rendererOpts source target =
    runResource sourceFormat rendererOpts
      (sourceHandle source)
      (sinkFile target)

-- | Transform from a file to lazy 'TL.Text'
--
-- @since 0.0.1.0
transformFileToText
  :: MonadUnliftIO m
  => SourceFormat
  -> Renderer.Options
  -> FilePath
  -> m TL.Text
transformFileToText sourceFormat rendererOpts source =
    runResource sourceFormat rendererOpts (sourceFile source) sinkLazyText

-- | Transform from a file to a 'Handle'
--
-- @since 0.0.1.0
transformFileToHandle
  :: MonadUnliftIO m
  => SourceFormat
  -> Renderer.Options
  -> FilePath
  -> Handle
  -> m ()
transformFileToHandle sourceFormat rendererOpts source target =
    runResource sourceFormat rendererOpts
      (sourceFile source)
      (sinkHandle target)

-- | Transform from a file to a file
--
-- @since 0.0.1.0
transformFileToFile
  :: MonadUnliftIO m
  => SourceFormat
  -> Renderer.Options
  -> FilePath
  -> FilePath
  -> m ()
transformFileToFile sourceFormat rendererOpts source target =
    runResource sourceFormat rendererOpts
      (sourceFile source)
      (sinkFile target)

------------------------------------------------------------------------------
-- $LowLevelAPI
--
-- This low-level API provides more control over transforming literate source
-- code, using "Conduit".  The 'transform' transformer implements the
-- transformation, transforming lines of input 'T.Text' to lines of output
-- 'T.Text'.  Various producers are provided to produce lines of input
-- 'T.Text' from common sources, and various consumers are provided to consume
-- lines of output 'T.Text' to common sinks.  These can be used separately if
-- necessary, but some run functions are provided for common usage.
--
-- Example usage:
--
-- @
-- {-\# LANGUAGE OverloadedStrings \#-}
--
-- module Main (main) where
--
-- -- https://hackage.haskell.org/package/base
-- import System.IO (stdout)
--
-- -- https://hackage.haskell.org/package/literatex
-- import qualified LiterateX
-- import qualified LiterateX.Renderer as Renderer
-- import qualified LiterateX.Types.SourceFormat as SourceFormat
--
-- main :: IO ()
-- main = do
--     let demoBS = "..."
--     LiterateX.runIO
--       SourceFormat.LiterateHaskell
--       (Renderer.defaultOptionsFor "haskell")
--       (LiterateX.sourceByteString demoBS)
--       (LiterateX.sinkHandle stdout)
-- @

-- | Run a pure LiterateX transformation
--
-- This function works with the following input line producers:
--
-- * 'sourceString'
-- * 'sourceText'
-- * 'sourceLazyText'
-- * 'sourceByteString'
-- * 'sourceLazyByteString'
--
-- This function works with the following output line consumers:
--
-- * 'sinkString'
-- * 'sinkText'
-- * 'sinkLazyText'
-- * 'sinkByteString'
-- * 'sinkLazyByteString'
--
-- @since 0.0.1.0
runPure
  :: SourceFormat
  -> Renderer.Options
  -> C.ConduitT () T.Text C.Identity ()     -- ^ input line producer
  -> C.ConduitT T.Text C.Void C.Identity r  -- ^ output line consumer
  -> r
runPure sourceFormat rendererOpts source sink = C.runConduitPure $
    source .| transform sourceFormat rendererOpts .| sink

-- | Run a LiterateX transformation using IO
--
-- This function works with the following input line producers:
--
-- * 'sourceString'
-- * 'sourceText'
-- * 'sourceLazyText'
-- * 'sourceByteString'
-- * 'sourceLazyByteString'
-- * 'LiterateX.sourceHandle'
--
-- This function works with the following output line consumers:
--
-- * 'sinkString'
-- * 'sinkText'
-- * 'sinkLazyText'
-- * 'sinkByteString'
-- * 'sinkLazyByteString'
-- * 'LiterateX.sinkHandle'
--
-- @since 0.0.1.0
runIO
  :: MonadIO m
  => SourceFormat
  -> Renderer.Options
  -> C.ConduitT () T.Text m ()     -- ^ input line producer
  -> C.ConduitT T.Text C.Void m r  -- ^ output line consumer
  -> m r
runIO sourceFormat rendererOpts source sink = C.runConduit $
    source .| transform sourceFormat rendererOpts .| sink

-- | Run a LiterateX transformation using resource management
--
-- This function works with the following input line producers:
--
-- * 'sourceString'
-- * 'sourceText'
-- * 'sourceLazyText'
-- * 'sourceByteString'
-- * 'sourceLazyByteString'
-- * 'LiterateX.sourceHandle'
-- * 'LiterateX.sourceFile'
--
-- This function works with the following output line consumers:
--
-- * 'sinkString'
-- * 'sinkText'
-- * 'sinkLazyText'
-- * 'sinkByteString'
-- * 'sinkLazyByteString'
-- * 'LiterateX.sinkHandle'
-- * 'LiterateX.sinkFile'
--
-- @since 0.0.1.0
runResource
  :: MonadUnliftIO m
  => SourceFormat
  -> Renderer.Options
  -> C.ConduitT () T.Text (C.ResourceT m) ()     -- ^ input line producer
  -> C.ConduitT T.Text C.Void (C.ResourceT m) r  -- ^ output line consumer
  -> m r
runResource sourceFormat rendererOpts source sink = C.runConduitRes $
    source .| transform sourceFormat rendererOpts .| sink

------------------------------------------------------------------------------
-- $Producers

-- | Produce input lines from a 'String' source
--
-- @since 0.0.1.0
sourceString
  :: Monad m
  => String
  -> C.ConduitT i T.Text m ()
sourceString = CC.yieldMany . map T.pack . lines

-- | Produce input lines from a 'T.Text' source
--
-- @since 0.0.1.0
sourceText
  :: Monad m
  => T.Text
  -> C.ConduitT i T.Text m ()
sourceText source = C.yield source .| CC.linesUnbounded

-- | Produce input lines from a lazy 'TL.Text' source
--
-- @since 0.0.1.0
sourceLazyText
  :: Monad m
  => TL.Text
  -> C.ConduitT i T.Text m ()
sourceLazyText source = CC.sourceLazy source .| CC.linesUnbounded

-- | Produce input lines from a 'BS.ByteString' source
--
-- @since 0.0.1.0
sourceByteString
  :: Monad m
  => BS.ByteString
  -> C.ConduitT i T.Text m ()
sourceByteString source =
    C.yield source .| CC.linesUnboundedAscii .| decodeUtf8LinesLenient

-- | Produce input lines from a lazy 'BS.ByteString' source
--
-- @since 0.0.1.0
sourceLazyByteString
  :: Monad m
  => BSL.ByteString
  -> C.ConduitT i T.Text m ()
sourceLazyByteString source =
    CC.sourceLazy source .| CC.linesUnboundedAscii .| decodeUtf8LinesLenient

-- | Produce input lines from a 'Handle' source
--
-- @since 0.0.1.0
sourceHandle
  :: MonadIO m
  => Handle
  -> C.ConduitT i T.Text m ()
sourceHandle source =
    CC.sourceHandle source .| CC.linesUnboundedAscii .| decodeUtf8LinesLenient

-- | Produce input lines from a file source
--
-- @since 0.0.1.0
sourceFile
  :: C.MonadResource m
  => FilePath
  -> C.ConduitT i T.Text m ()
sourceFile source =
    CC.sourceFile source .| CC.linesUnboundedAscii .| decodeUtf8LinesLenient

------------------------------------------------------------------------------
-- $Consumers

-- | Consume output lines, returning a 'String'
--
-- @since 0.0.1.0
sinkString
  :: Monad m
  => C.ConduitT T.Text o m String
sinkString = fmap TL.unpack $ CC.unlines .| CC.sinkLazy

-- | Consume output lines, returning 'T.Text'
--
-- @since 0.0.1.0
sinkText
  :: Monad m
  => C.ConduitT T.Text o m T.Text
sinkText = fmap TL.toStrict $ CC.unlines .| CC.sinkLazy

-- | Consume output lines, returning lazy 'T.Text'
--
-- @since 0.0.1.0
sinkLazyText
  :: Monad m
  => C.ConduitT T.Text o m TL.Text
sinkLazyText = CC.unlines .| CC.sinkLazy

-- | Consume output lines, returning a 'BS.ByteString'
--
-- @since 0.0.1.0
sinkByteString
  :: Monad m
  => C.ConduitT T.Text o m BS.ByteString
sinkByteString =
    fmap BSL.toStrict $ CC.unlines .| CC.encodeUtf8 .| CC.sinkLazy

-- | Consume output lines, returning a lazy 'BS.ByteString'
--
-- @since 0.0.1.0
sinkLazyByteString
  :: Monad m
  => C.ConduitT T.Text o m BSL.ByteString
sinkLazyByteString = CC.unlines .| CC.encodeUtf8 .| CC.sinkLazy

-- | Consume output lines, writing to a 'Handle'
--
-- @since 0.0.1.0
sinkHandle
  :: MonadIO m
  => Handle
  -> C.ConduitT T.Text o m ()
sinkHandle target = CC.encodeUtf8 .| CC.unlinesAscii .| CC.sinkHandle target

-- | Consume output lines, writing to a file
--
-- @since 0.0.1.0
sinkFile
  :: C.MonadResource m
  => FilePath
  -> C.ConduitT T.Text o m ()
sinkFile target = CC.encodeUtf8 .| CC.unlinesAscii .| CC.sinkFile target

------------------------------------------------------------------------------
-- $Transformers

-- | Transform input lines to output lines
--
-- @since 0.0.1.0
transform
  :: Monad m
  => SourceFormat
  -> Renderer.Options
  -> C.ConduitT T.Text T.Text m ()
transform sourceFormat rendererOpts
    = Parser.parse sourceFormat
    .| CL.concatMapAccum (\x n -> (n + 1, [(n, x)])) 1
    .| Renderer.render rendererOpts

------------------------------------------------------------------------------
-- $Internal

-- | Decode UTF-8 'BS.ByteString' lines to 'T.Text'
--
-- Note that 'CC.decodeUtf8Lenient' decodes UTF-8 streams and therefore
-- discards empty strings.  This version decodes UTF-8 lines and does not
-- discard empty lines.
decodeUtf8LinesLenient
  :: Monad m
  => C.ConduitT BS.ByteString T.Text m ()
decodeUtf8LinesLenient =
    C.awaitForever $ C.yield . TE.decodeUtf8With TEE.lenientDecode
